In [3]:
import pandas as pd

In [5]:
df = pd.read_excel("raw_dataset.xlsx")

In [ ]:
df.head()

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


# 1. Dataset Overview

Before cleaning the dataset, we first understand its structure, size, columns, and data types.

In [ ]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 1200
Number of columns: 14


In [ ]:
print("Column Names:")
for column in df.columns:
    print("-", column)

Column Names:
- OrderID
- Date
- CustomerID
- Product
- Quantity
- UnitPrice
- ShippingAddress
- PaymentMethod
- OrderStatus
- TrackingNumber
- ItemsInCart
- CouponCode
- ReferralSource
- TotalPrice


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   OrderID          1200 non-null   object        
 1   Date             1200 non-null   datetime64[ns]
 2   CustomerID       1200 non-null   object        
 3   Product          1200 non-null   object        
 4   Quantity         1200 non-null   int64         
 5   UnitPrice        1200 non-null   float64       
 6   ShippingAddress  1200 non-null   object        
 7   PaymentMethod    1200 non-null   object        
 8   OrderStatus      1200 non-null   object        
 9   TrackingNumber   1200 non-null   object        
 10  ItemsInCart      1200 non-null   int64         
 11  CouponCode       891 non-null    object        
 12  ReferralSource   1200 non-null   object        
 13  TotalPrice       1200 non-null   float64       
dtypes: datetime64[ns](1), float64(2), int64(

In [ ]:
missing_values = df.isnull().sum()

print("Missing values in each column:")
print(missing_values)

Missing values in each column:
OrderID              0
Date                 0
CustomerID           0
Product              0
Quantity             0
UnitPrice            0
ShippingAddress      0
PaymentMethod        0
OrderStatus          0
TrackingNumber       0
ItemsInCart          0
CouponCode         309
ReferralSource       0
TotalPrice           0
dtype: int64


In [ ]:
duplicate_rows = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_rows)

Number of duplicate rows: 0


In [ ]:
duplicate_order_ids = df["OrderID"].duplicated().sum()

print("Number of duplicate OrderIDs:", duplicate_order_ids)

Number of duplicate OrderIDs: 0


In [ ]:
print("Date data type:", df["Date"].dtype)

Date data type: datetime64[ns]


In [ ]:
date_check = pd.to_datetime(df["Date"], errors="coerce")

invalid_dates = date_check.isna().sum()

print("Number of invalid dates:", invalid_dates)

Number of invalid dates: 0


In [ ]:
print("Earliest date:", df["Date"].min())
print("Latest date:", df["Date"].max())

Earliest date: 2023-01-01 00:00:00
Latest date: 2025-06-30 00:00:00


In [ ]:
numeric_columns = ["Quantity", "UnitPrice", "ItemsInCart", "TotalPrice"]

print(df[numeric_columns].dtypes)

Quantity         int64
UnitPrice      float64
ItemsInCart      int64
TotalPrice     float64
dtype: object


In [ ]:
for column in numeric_columns:
    numeric_values = pd.to_numeric(df[column], errors="coerce")
    invalid_count = numeric_values.isna().sum()
    print(column, "→ invalid numeric values:", invalid_count)

Quantity → invalid numeric values: 0
UnitPrice → invalid numeric values: 0
ItemsInCart → invalid numeric values: 0
TotalPrice → invalid numeric values: 0


In [6]:
text_columns = [
    "Product",
    "PaymentMethod",
    "OrderStatus",
    "CouponCode",
    "ReferralSource"
]

for column in text_columns:
    print("\n", column)
    print(df[column].dropna().unique())


 Product
['Monitor' 'Phone' 'Tablet' 'Chair' 'Printer' 'Laptop' 'Desk']

 PaymentMethod
['Debit Card' 'Online' 'Credit Card' 'Gift Card' 'Cash']

 OrderStatus
['Shipped' 'Cancelled' 'Returned' 'Delivered' 'Pending']

 CouponCode
['SAVE10' 'FREESHIP' 'WINTER15']

 ReferralSource
['Instagram' 'Referral' 'Email' 'Facebook' 'Google']


## 2. Handling Missing Values

The `CouponCode` column contains missing values. Since a missing coupon code indicates that no coupon was provided or used, these records should not be deleted.

The missing values will be replaced with `NO_COUPON` to make the information explicit while preserving all order records.

In [7]:
print("Missing CouponCode values before cleaning:",
      df["CouponCode"].isna().sum())

Missing CouponCode values before cleaning: 309


In [9]:
df["CouponCode"] = df["CouponCode"].fillna("NO_COUPON")

In [10]:
print("Missing values after cleaning:")
print(df.isnull().sum())

Missing values after cleaning:
OrderID            0
Date               0
CustomerID         0
Product            0
Quantity           0
UnitPrice          0
ShippingAddress    0
PaymentMethod      0
OrderStatus        0
TrackingNumber     0
ItemsInCart        0
CouponCode         0
ReferralSource     0
TotalPrice         0
dtype: int64


In [11]:
print("Duplicate rows after cleaning:", df.duplicated().sum())

Duplicate rows after cleaning: 0


In [12]:
print("Duplicate OrderIDs after cleaning:",
      df["OrderID"].duplicated().sum())

Duplicate OrderIDs after cleaning: 0


In [13]:
date_check = pd.to_datetime(df["Date"], errors="coerce")

print("Invalid dates after cleaning:", date_check.isna().sum())

Invalid dates after cleaning: 0


In [14]:
numeric_columns = ["Quantity", "UnitPrice", "ItemsInCart", "TotalPrice"]

for column in numeric_columns:
    invalid = pd.to_numeric(df[column], errors="coerce").isna().sum()
    print(column, "→ invalid values:", invalid)

Quantity → invalid values: 0
UnitPrice → invalid values: 0
ItemsInCart → invalid values: 0
TotalPrice → invalid values: 0


In [15]:
print("Total missing values in cleaned dataset:",
      df.isnull().sum().sum())

Total missing values in cleaned dataset: 0


In [16]:
# STEP 26A - Export Final Cleaned Dataset

cleaned_file = "cleaned_dataset.xlsx"

df.to_excel(cleaned_file, index=False)

print("Final cleaned Excel file created successfully!")
print("File name:", cleaned_file)
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Final cleaned Excel file created successfully!
File name: cleaned_dataset.xlsx
Number of rows: 1200
Number of columns: 14


# Data Cleaning & Preparation Report

## 1. Project Overview

This project focuses on cleaning and preparing a raw dataset for further data analysis.

The main objectives were to identify and handle missing values, check for duplicate records, validate dates and numeric values, and ensure that the dataset was ready for analysis.

---

## 2. Dataset Summary

| Item | Result |
|---|---:|
| Total Rows | 1,200 |
| Total Columns | 14 |
| Duplicate Rows | 0 |
| Duplicate OrderIDs | 0 |
| Invalid Dates | 0 |
| Invalid Numeric Values | 0 |
| Missing Values After Cleaning | 0 |

---

## 3. Missing Value Handling

The `CouponCode` column contained 309 missing values.

Instead of deleting these records, the missing values were replaced with:

`NO_COUPON`

This indicates that no coupon code was recorded for those orders while preserving all dataset records.

---

## 4. Duplicate Record Check

The dataset was checked for duplicate rows and duplicate `OrderID` values.

Results:

- Duplicate rows: 0
- Duplicate OrderIDs: 0

Therefore, the dataset contains no duplicate records or duplicate order identifiers.

---

## 5. Date Validation

The `Date` column was checked for incorrectly formatted or invalid dates.

Invalid dates found:

`0`

Therefore, all dates were successfully validated.

---

## 6. Numeric Data Validation

The following numeric columns were checked:

- Quantity
- UnitPrice
- ItemsInCart
- TotalPrice

No invalid numeric values were found.

---

## 7. Data Consistency Check

The `TotalPrice` column was checked against:

`Quantity × UnitPrice`

All records were consistent with the expected calculation.

---

## 8. Final Result

After cleaning and validation, the dataset contains:

- 1,200 valid records
- 14 columns
- No duplicate rows
- No duplicate OrderIDs
- No invalid dates
- No invalid numeric values
- No remaining missing values

The cleaned dataset is ready for further data analysis and visualization.

---

## 9. Output File

Final cleaned dataset:

`cleaned_dataset.xlsx`